### PR TITLE
Add a header to all serialised data & switch to compatibility serializer.

### PR DESCRIPTION
**Pull request by rick-r3**
1. Add a header so we quickly detect attempts to deserialize the wrong data and embed a versioning scheme to allow evolution of our serialization format.  Currently the version number is major.minor where major is a single byte integer and minor is two-byte integer.
2. Switch to the compatibility serializer to allow additional resilience to changes to classes.  The additional option to allow fields overridden in subclasses was required for tests to pass.

### CODE REVIEW

**Comment by rick-r3** [comment]
Note this requires PR #267 to be merged to master first.

**Comment by mikehearn** [comment]
LGTM

File: core/src/main/kotlin/net/corda/core/serialization/Kryo.kt [hunk 1]
```diff
@@ -79,10 +79,19 @@ class SerializedBytes<T : Any>(bytes: ByteArray, val internalOnly: Boolean = fal
...
+// "corda" + majorVersionByte + minorVersionMSB + minorVersionLSB
+private val KryoHeaderV0_1: OpaqueBytes = OpaqueBytes("corda\u0000\u0000\u0001".toByteArray())
+
 // Some extension functions that make deserialisation convenient and provide auto-casting of the result.
 fun <T : Any> ByteArray.deserialize(kryo: Kryo = threadLocalP2PKryo()): T {
-    @Suppress("UNCHECKED_CAST")
-    return kryo.readClassAndObject(Input(this)) as T
+    Input(this).use {
+        val header = OpaqueBytes(it.readBytes(8))
+        if (header != KryoHeaderV0_1) {
```
Comment (exFalso) [inline core/src/main/kotlin/net/corda/core/serialization/Kryo.kt]: What is the upgrade policy of the version number? We're doing a hard equality check here, so I'm assuming minor version changes also imply API breaks? Does this mean that changes that play well with `CompatibleFieldSerializer` will not bump the version?

Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/serialization/Kryo.kt]: My general thinking was to have major version bump when we change scheme.  Minor version bump would perhaps be equivalent to changing the custom serialiser for a particular class that makes it binary incompatible.  If it's actually handled by the compatibility serialiser for us, then there's no need to change the header version.

Generally changing the serialisation such that the header needs to change will require some serious thought with respect to deployment.

Obviously right now I've gone for the bare bones support where the deserialisation code assumes the header is equal or not.

**Pull request merged**